dbt/include/spark/macros/materializations/table.sql

```diff
@@ -1,25 +1,64 @@
-{% materialization table, adapter = 'spark' %}
-  
+{% materialization table, adapter='spark' %}
   {%- set identifier = model['alias'] -%}
+  {%- set tmp_identifier = model['name'] + '__dbt_tmp' -%}
+  {%- set backup_identifier = model['name'] + '__dbt_backup' -%}
 
   {%- set old_relation = adapter.get_relation(database=database, schema=schema, identifier=identifier) -%}
   {%- set target_relation = api.Relation.create(identifier=identifier,
                                                 schema=schema,
                                                 database=database,
                                                 type='table') -%}
+  {%- set intermediate_relation = api.Relation.create(identifier=tmp_identifier,
+                                                      schema=schema,
+                                                      database=database,
+                                                      type='table') -%}
 
-  {{ run_hooks(pre_hooks) }}
+  /*
+      See ../view/view.sql for more information about this relation.
+  */
+  {%- set backup_relation_type = 'table' if old_relation is none else get_relation_type(old_relation) -%}
+  {%- set backup_relation = api.Relation.create(identifier=backup_identifier,
+                                                schema=schema,
+                                                database=database,
+                                                type=backup_relation_type) -%}
+
+  {%- set exists_as_view = (old_relation is not none and old_relation.is_view) -%}
 
-  -- setup: if the target relation already exists, drop it
-  {% if old_relation -%}
+  -- drop the temp relations if they exist for some reason
+  {{ adapter.drop_relation(intermediate_relation) }}
+  {{ adapter.drop_relation(backup_relation) }}
+
+  {% if exists_as_view -%}
     {{ adapter.drop_relation(old_relation) }}
+    {%- set old_relation = none -%}
   {%- endif %}
 
+  {{ run_hooks(pre_hooks, inside_transaction=False) }}
+
+  -- `BEGIN` happens here:
+  {{ run_hooks(pre_hooks, inside_transaction=True) }}
+
   -- build model
   {% call statement('main') -%}
-    {{ create_table_as(False, target_relation, sql) }}
+    {{ create_table_as(False, intermediate_relation, sql) }}
```

Comment:
We switched (in #17) from using the default (Postgres/Redshift) approach (creating an intermediate table and performing a rename-swap) to a naive approach (drop the existing table, create it anew) for two main reasons:

1. The appeal of this clever approach on Postgres/Redshift is the atomicity of running the whole thing inside of a transaction (except for the final `drop`). Spark doesn't have transactions, so if a query fails, it gets stuck midway through the process. It felt better to have a simple/intuitive process of only two steps (drop + create) than potentially have `__dbt_tmp` and `__dbt_backup` relational objects hanging around.
2. In Databricks (and potentially other hosted Spark providers), if it's a managed table, renaming a table physically moves the contents of that table from one file location to another. We found that the `alter ... rename` step can take _longer_ than the `create` step.

I'd be open to hearing arguments in favor of returning to this default, cleverer approach.

Reply:
@jtcohen6 For drop + create approach, the table is not available before create complete. Right? Any suggestions if I want to use that for dim table? That table might be temporarily unavailable?

Reply:
> @jtcohen6 For drop + create approach, the table is not available before create complete. Right? Any suggestions if I want to use that for dim table? That table might be temporarily unavailable?

That's correct, but it's true for the default/Redshift/Postgres approach too (`create ... __dbt_tmp` + `alter ... rename` + `alter ... rename` + `drop`), because Spark doesn't have transactions. The table will be unavailable while the `alter ... rename` is occurring, and we found that many renames (especially for managed tables) take _longer_ than simply recreating the table from scratch. So the `drop` + `create` approach is both simpler and, in many cases, results in _less_ downtime of a dim table.

The only atomic operation currently available in Spark for dropping and recreating a table is `insert overwrite`. Though it feels a bit counterintuitive, you could materialize your dim table as an incremental model, and re-select all the data (= overwrite all partitions) every time.

I think the best answer is that `create or replace table` is coming in Spark 3.0.0.

Reply:
Thanks for your comments.